### DISCUSSION TITLE
How Linux works - from CPU to programs in high level languages

### DISCUSSION THREAD

**Discussion by marcinjahn** [elementary/edw#15] [Ideas]
Speaker: **TBD**

I'd love to see a talk which uncovers a bit of the "magic" behind our computers. I know that the title sounds like it could be a 1000 pages long book, but that is not the case. I'd see it as a presentation of layers that computers nowadays use, without going too much into detail about each of the layers, but saying enough that the audience gets the general idea of why this layer is even there.
Going through layers such as CPU (machine code) -> assembler (symbolic machine code) -> BIOS/UEFI -> bootloader -> OS -> high-level programs, and how these layers relate to each other is in general what I mean.

**Comment by alcinnz** [comment]
I could offer to give this talk! I enjoy understanding & explaining how software works!

Here's some of my ongoing writings: https://adrian.geek.nz/docs

**Comment by danirabbit** [comment]
I'm not quite sure this is right for the audience of this particular conference. The focus here is on providing information for app developers, and I'm not sure how this content is super relevant to helping them write better apps

**Reply by marcinjahn** [reply]
I agree that it does not help in any way in building typical apps. However, such knowledge broadens the horizon a bit. I think that a lot of develops do not really know what's going on below the layer they are working on. Certainly, most people have an idea that their code is compiled into something lower-level, but how does it get executed? How does it happen that a window with my app is being created on the screen?

Maybe I went a bit "too low" in the title of this proposed talk. Maybe going down to CPU is too much. What I'm thinking is: typical programmer will be able to develop an app for any operating system (let's focus on elementary though). However, only a part of them will be able to mess with something like window manager, or display manager. It's much more complex because it requires you to understand what is actually happening below the user-space apps level.

I think the talk that would uncover a bit of how the system is built, what are the components, what talks with what, etc., would be quite useful. Especially, if it was done with practical examples.

Again, the title and description that I provided were very broad. I think the focus should be limited to a subset of what I originally proposed. That way the talk could be enough detailed. If too big range of "layers" is taken into account, it will be too general (not sure what amount of time one presentation should take).

**Reply by alcinnz** [reply]
Yeah, it's an extremely broad topic which could fill several talks! Going into much detail would not fit even if it was relevant.

At the same time I think the ability to dig into this is part of what makes platforms like elementary OS so special, and worth highlighting. Besides it may be useful in case app developers need more from the OS, and (unlike the web) a useful way to understand the GNOME/elementary APIs.

**Comment by alcinnz** [comment]
Given how broad this topic is, I'm interested in discussing what it should focus on.

To me GTK, GLib, Vala (which I'd have to contrast to C), the FreeDesktop.Org specs, Gala/Mutter, DBus, & Linux should be the primary topics. The graphics and text stacks could also be relevant and should definitely be mentioned, but I'm guessing they won't fit.

I wouldn't mention the BIOS, for example, as that's not active past bootup.

**Comment by marbetschar** [comment]
I'd be interested in such a talk if it focuses on an overview of the core components of the elementary Desktop: Login Manager, Window Manager, X11 vs. Wayland, DBus and maybe keyboard or mouse I/O.

Which components are involved? What are their responsibilities? What are their names in elementary? How do they work together? How do they fit in the wider Linux ecosystem?

Basically a talk explaining how we go "From code to pixels and back again on a modern Linux Desktop Environment".
